dotnet/src/Connectors/Connectors.Memory.PgVector/PostgresServiceCollectionExtensions.cs

```diff
@@ -1,163 +1,295 @@
+            static (sp) => sp.GetRequiredService<PostgresVectorStore>(), lifetime));
 
         return services;
     }
 
     /// <summary>
-    /// Register a Postgres <see cref="VectorStore"/> with the specified service ID and where an NpgsqlDataSource is constructed using the provided parameters.
+    /// Registers a <see cref="PostgresVectorStore"/> as <see cref="VectorStore"/>, with the specified connection string and service lifetime.
+    /// </summary>
+    /// <inheritdoc cref="AddKeyedPostgresVectorStore"/>
+    public static IServiceCollection AddPostgresVectorStore(
+        this IServiceCollection services,
+        string connectionString,
+        PostgresVectorStoreOptions? options = default,
+        ServiceLifetime lifetime = ServiceLifetime.Singleton)
+    {
+        Verify.NotNullOrWhiteSpace(connectionString);
+
+        return AddVectorStore(services, serviceKey: null, sp => connectionString, sp => options, lifetime);
+    }
+
+    /// <summary>
+    /// Registers a keyed <see cref="PostgresVectorStore"/> as <see cref="VectorStore"/>, with the specified connection string and service lifetime.
     /// </summary>
     /// <param name="services">The <see cref="IServiceCollection"/> to register the <see cref="VectorStore"/> on.</param>
+    /// <param name="serviceKey">The key with which to associate the vector store.</param>
     /// <param name="connectionString">Postgres database connection string.</param>
     /// <param name="options">Optional options to further configure the <see cref="VectorStore"/>.</param>
-    /// <param name="serviceId">An optional service id to use as the service key.</param>
+    /// <param name="lifetime">The service lifetime for the store. Defaults to <see cref="ServiceLifetime.Singleton"/>.</param>
     /// <returns>The service collection.</returns>
-    public static IServiceCollection AddPostgresVectorStore(this IServiceCollection services, string connectionString, PostgresVectorStoreOptions? options = default, string? serviceId = default)
+    public static IServiceCollection AddKeyedPostgresVectorStore(
+        this IServiceCollection services,
+        object serviceKey,
+        string connectionString,
+        PostgresVectorStoreOptions? options = default,
+        ServiceLifetime lifetime = ServiceLifetime.Singleton)
     {
-        services.AddKeyedSingleton<VectorStore>(
-            serviceId,
-            (sp, obj) =>
-            {
-                options ??= sp.GetService<PostgresVectorStoreOptions>() ?? new()
-                {
-                    EmbeddingGenerator = sp.GetService<IEmbeddingGenerator>()
-                };
-
-                return new PostgresVectorStore(connectionString, options);
-            });
+        Verify.NotNull(serviceKey);
+        Verify.NotNullOrWhiteSpace(connectionString);
+
+        return AddVectorStore(services, serviceKey, sp => connectionString, sp => options, lifetime);
+    }
+
+    private static IServiceCollection AddVectorStore(
```

Comment:
So we don't want to expose a public AddPostgresVectorStore that accepts Funcs (like we do for AddPostgresCollection)? Registering stores (instead of collections) is indeed less important, but I can see users complaining here.

(but in any case, am OK to start like this and add more overloads later based on user request).

Reply:
`AddPostgresVectorStore` is expected to be used least frequently. For all the providers, I wanted to have only the `Func` overload as it allows for anything (reading from config, hardcoded values etc). But Postgres is having plenty of downloads, so I wanted to lower the chance of breaking the users by changing the methods.

Having said that, I am just going to add it for Postgres.